### DISCUSSION TITLE
Can it only be used to process label data?

### DISCUSSION THREAD

**Discussion by CDPDisk** [seung-lab/connected-components-3d#79] [Ideas]
Can it handle continuous data? For example, connect voxels with similar values.
I saw a similar function on SimpleITK, which uses connected components to segment lungs or other organs with close values.

**Comment by william-silversmith** [comment]
This library isn't meant to perform the segmentation by itself. I can see how integrating CCL into a segmentation algorithm could possibly improve overall performance, but the operations are fundamentally different. This CCL library operates on pre-segmented data in order to avoid introducing a dependency on a particular segmentation algorithm (which can range from a simple threshold to a complex neural net).

**Reply by CDPDisk** [reply]
Thank you for your answer. I reviewed the lung segmentation. It does not use connected components, but uses thresholds, binary opening and closing.

My original idea was that if for the current calculated voxel, that is, the variable `cur`, can the surrounding voxels (such as 26 voxels) be connected when their values are similar
For example, when considering N and E voxels, `cur` and `in_labels` are float type data:
```
if (z> 0 && cur-in_labels[loc + E] <0.5) {
           out_labels[loc] = out_labels[loc + E];
         }
```
In this way, even when the data is floating point, it can be connected and calculated. So its function looks like the Magic Wand Tool of photoshop, hahaha🤣.
But I did find it seems useless

**Reply by william-silversmith** [reply]
That's pretty interesting though! Do you have some samples of before and after you could show?

**Reply by CDPDisk** [reply]
I found an example. There are some squares in this picture, but obviously the pixel values in the same square is not equal, so it is not possible to directly use this data as an int64 to calculate connected-components. And by connecting pixels with similar pixel values, the effect of the second image can be achieved
![1](https://user-images.githubusercontent.com/32790144/142581836-3477a78a-1766-4e76-9503-f36a636f3354.png)
![2](https://user-images.githubusercontent.com/32790144/142581839-27be21bd-8f42-493a-9531-59b0095db067.png)

An other example is that split black background which the value is 0 in the CT or MR image. But there always be some noise in the background , so it is hard to segment it by `background = data[data==0]`. But I think it can be easy using the connected components in the float data.

It just an idea. I think it is unnecessary in most cases.

**Reply by william-silversmith** [reply]
This is an interesting technique and I'll keep it in mind! I'd be open to adding this functionality, possibly as an extra parameter `delta=0.5` with the default value being zero. Is this something you need in the near future? If so, I'm happy to take a PR that has two code paths (or a single code path that is validated to be as fast as the original within 5%). I'll consider doing it myself, but I'm just a bit busy at the moment.

**Reply by CDPDisk** [reply]
I don't need it in the near future, but I would love to see this feature appear, haha. Thank you for your approval, but my coding skills are not good😢
I look forward to seeing it completed someday, haha.

**Reply by william-silversmith** [reply]
I gave it a try. It looks like the magic wand version needs a different decision tree because certain assumptions are violated. There was also a 20% performance penalty on 8-connected just replacing the comparison operators. I think this needs to be a separate file with a re-implementation of each operator. Will give that a try some other time!

**Comment by william-silversmith** [comment]
Something like this? :D I made a quick prototype. Image is from https://en.wikipedia.org/wiki/Scanning_electron_microscope#/media/File:Misc_pollen.jpg

<img width="769" alt="image" src="https://user-images.githubusercontent.com/2517065/142824941-1bb14193-cea3-4aad-82fb-79df71eda541.png">

Code + I clicked on a few segments.

```python
import numpy as np
import cc3d
from cloudvolume import hyperview, view
from PIL import Image

img = Image.open("Misc_pollen.jpg")
img = np.array(img)[:,:,0]
img = (img > 55) * img

out = cc3d.connected_components(img, connectivity=8, delta=10)

hyperview(img.T, out.T)
```

**Reply by william-silversmith** [reply]
You can see the code here: https://github.com/seung-lab/connected-components-3d/pull/80

**Reply by CDPDisk** [reply]
Great! Thanks your work!

**Reply by william-silversmith** [reply]
Version 3.4.0 now has this feature available for download. Enjoy!
